File: ui/src/lib/api/presets.ts
import { request } from "./client";

// OpenRag pipeline-preset registry (Phase 14). Mounted at `/presets` (admin-only).
// Shapes verified against openrag/api/schemas/admin/preset_schemas.py + routers/admin/presets.py:
//   GET    /presets/options              { chunking_strategies, retrieval_types, reranker_providers }
//   GET    /presets/                     list (bare array; optional ?preset_type=)
//   GET    /presets/{type}/{name}        get one
//   POST   /presets/                     create → 201
//   PUT    /presets/{type}/{name}        update
//   DELETE /presets/{type}/{name}        delete → 204

export type PresetType = "indexation" | "retrieval";

export interface PresetResponse {
  name: string;
  preset_type: PresetType;
  config: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  // Distinct partition count referencing this preset. Only populated by
  // listPresets(); get/create/update responses default this to 0.
  used_by_partitions: number;
}

export interface CreatePresetRequest {
  name: string;
  preset_type: PresetType;
  config?: Record<string, unknown>;
}

export interface UpdatePresetRequest {
  name?: string;
  config?: Record<string, unknown>;
}

export interface PresetOptionsResponse {
  chunking_strategies: string[];
  // Optional: older backends don't return this (UI falls back to a default list).
  parsing_strategies?: string[];
  retrieval_types: string[];
  reranker_providers: string[];
  // Optional: older backends predate context compression.
  compression_available?: boolean;
  compression_backend?: string;
}

const BASE = "/presets";
const enc = encodeURIComponent;

export function getPresetOptions() {
  return request<PresetOptionsResponse>(`${BASE}/options`);
}

/** List presets (bare array). Optionally filter by preset type. */
export function listPresets(presetType?: PresetType) {
  const qs = presetType ? `?preset_type=${enc(presetType)}` : "";
  return request<PresetResponse[]>(`${BASE}/${qs}`);
}

export function getPreset(presetType: PresetType, name: string) {
  return request<PresetResponse>(`${BASE}/${enc(presetType)}/${enc(name)}`);
}

export function createPreset(data: CreatePresetRequest) {
  return request<PresetResponse>(`${BASE}/`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function updatePreset(
  presetType: PresetType,
  name: string,
  data: UpdatePresetRequest,
) {
  return request<PresetResponse>(`${BASE}/${enc(presetType)}/${enc(name)}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export function deletePreset(presetType: PresetType, name: string) {
  return request<void>(`${BASE}/${enc(presetType)}/${enc(name)}`, {
    method: "DELETE",
  });
}

File: openrag/api/routers/admin/presets.py
"""Admin routes for the Phase 14 pipeline preset registry."""

from api.dependencies.auth import require_admin
from api.schemas.admin.preset_schemas import (
    CreatePresetRequest,
    PresetOptionsResponse,
    PresetResponse,
    PresetType,
    UpdatePresetRequest,
)
from core.chunking import chunking_registry
from core.config.indexation_pipeline import PARSING_STRATEGIES
from core.config.table_reconstruction import TABLE_RECONSTRUCTION_MODES
from core.rerankers.registry import reranker_registry
from core.retrieval import retriever_registry
from di.providers import get_preset_service
from fastapi import APIRouter, Depends, Response, status

router = APIRouter(dependencies=[Depends(require_admin)])

_DEFAULT_RERANKER_PROVIDERS = ["infinity", "openai", "tei"]

# The selectable PDF backends. Shares the IndexationPipelineConfig constant so
# the exposed options can never drift from what the model accepts. ``None``
# (inherit the global PDFLOADER) is the field default, not an explicit choice,
# so it is not surfaced here.
_PARSING_STRATEGIES = list(PARSING_STRATEGIES)


def _registered_or_default(registered: list[str], defaults: list[str]) -> list[str]:
    """Return registry values, falling back to known defaults before DI imports providers."""
    return registered or defaults


@router.get("/options", response_model=PresetOptionsResponse)
async def get_preset_options():
    """Return available preset strategy choices."""
    return PresetOptionsResponse(
        chunking_strategies=chunking_registry.list_registered(),
        parsing_strategies=_PARSING_STRATEGIES,
        table_reconstruction_modes=list(TABLE_RECONSTRUCTION_MODES),
        retrieval_types=retriever_registry.list_registered(),
        reranker_providers=_registered_or_default(
            reranker_registry.list_registered(),
            _DEFAULT_RERANKER_PROVIDERS,
        ),
    )


@router.post("/", response_model=PresetResponse, status_code=status.HTTP_201_CREATED)
async def create_preset(
    body: CreatePresetRequest,
    service=Depends(get_preset_service),
):
    """Create a named pipeline preset."""
    return await service.create_preset(**body.model_dump())


@router.get("/", response_model=list[PresetResponse])
async def list_presets(
    preset_type: PresetType | None = None,
    service=Depends(get_preset_service),
):
    """List pipeline presets, optionally filtered by type."""
    return await service.list_presets(preset_type=preset_type)


@router.get("/{preset_type}/{name}", response_model=PresetResponse)
async def get_preset(
    preset_type: PresetType,
    name: str,
    service=Depends(get_preset_service),
):
    """Return one pipeline preset."""
    return await service.get_preset(name=name, preset_type=preset_type)


@router.put("/{preset_type}/{name}", response_model=PresetResponse)
async def update_preset(
    preset_type: PresetType,
    name: str,
    body: UpdatePresetRequest,
    service=Depends(get_preset_service),
):
    """Update a pipeline preset."""
    fields = body.model_dump(exclude_unset=True)
    if "name" in fields:
        fields["new_name"] = fields.pop("name")
    return await service.update_preset(
        name=name,
        preset_type=preset_type,
        **fields,
    )


@router.delete("/{preset_type}/{name}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_preset(
    preset_type: PresetType,
    name: str,
    service=Depends(get_preset_service),
):
    """Delete a pipeline preset."""
    await service.delete_preset(name=name, preset_type=preset_type)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
